### PR TITLE
gshrink

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

**Comment by phadej** [comment]
ping @kosmikus, can we merge and release this?

There is pending PR https://github.com/nick8325/quickcheck/pull/112, but until that, this functionality is practically useful!

_EDIT_ it seems QuickCheck has `genericsShrink` already

**Comment by kosmikus** [comment]
Sorry, I somehow didn't see this. I'll take a look.

File: src/Generics/SOP/Arbitrary.hs [hunk 1]
```diff
@@ -25,13 +25,41 @@ import Generics.SOP
...
 -- values of a particular size, and it is possible that this function
 -- diverges for recursive structures.
 --
 garbitrary :: forall a. (Generic a, All2 Arbitrary (Code a)) => Gen a
-garbitrary = liftM to $ hsequence =<< elements (apInjs_POP $ hcpure p arbitrary)
+garbitrary = fmap to $ oneof (map hsequence $ apInjs_POP $ hcpure p arbitrary)
   where
     p :: Proxy Arbitrary
     p = Proxy
+-- Note: we could generate all elements of sums and pick on with 'elements'.
```
Comment (kosmikus) [inline src/Generics/SOP/Arbitrary.hs]: The old version doesn't do what you seem to be implying here:

```
apInjs_POP $ hcpure p arbitrary :: [SOP Gen (Code a)]
elements (...) :: Gen (SOP Gen (Code a))
hsequence =<< ... :: Gen (SOP I (Code a))
```

Your version does:

```
apInjs_POP $ hcpure p arbitrary :: [SOP Gen (Code a)]
map hsequence $ ... :: [Gen (SOP I (Code a))]
oneof (...) :: Gen (SOP I (Code a))
```

I don't see why one would be significantly better than the other.

File: src/Generics/SOP/Arbitrary.hs [hunk 2]
```diff
@@ -25,13 +25,41 @@ import Generics.SOP
...
+    . hcmap (Proxy :: Proxy (All Arbitrary)) gshrink'
+    . unSOP
+    . from
+  where
+    gshrink' :: forall ys. All Arbitrary ys => NP I ys -> ([] :.: NP I) ys
+    gshrink' Nil          = Comp $ []
+    gshrink' (I x :* Nil) = Comp $ fmap (\x' -> I x' :* Nil) $ shrink x
+    gshrink' (I x :* xs)  = Comp $
+        [ I x' :* xs  | x'  <- shrink x ] ++
+        [ I x  :* xs' | xs' <- unComp $ gshrink' xs ]
```
Comment (kosmikus) [inline src/Generics/SOP/Arbitrary.hs:65]: It's not completely clear to me what the idea here is.

**Comment by kosmikus** [comment]
I'm certainly ok with the relaxed bounds, the gitignore change, the CI change and the Haddock fix. The QuickCheck stuff needs a bit more discussion.

File: src/Generics/SOP/Arbitrary.hs (hunk 1 above)
Comment (phadej) [reply src/Generics/SOP/Arbitrary.hs]: I see, I somehow though that we get `SOP Gen (Code a) -> Gen (SOP I (Code a))` with `hsequence`. I'll revert the change and add comments in between.

File: src/Generics/SOP/Arbitrary.hs (hunk 2 above)
Comment (phadej) [reply src/Generics/SOP/Arbitrary.hs:65]: The tuple instance hopefully clarifies:

``` hs
instance (Arbitrary a, Arbitrary b)
      => Arbitrary (a,b)
 where
  arbitrary = liftM2 (,) arbitrary arbitrary

  shrink (x, y) =
       [ (x', y) | x' <- shrink x ]
    ++ [ (x, y') | y' <- shrink y ]
```

E.g. we keep one component fixed and shrink another one. And then we do the same, reversed.

---

And now as I think properly, the singleton case isn't necessary

**Comment by phadej** [comment]
Reverted the `garbitrary` change. I used `do` notation, to clarify what happens there. I guess it'll be more useful for people seeking for examples of `generics-sop` usage.

File: src/Generics/SOP/Arbitrary.hs (hunk 2 above)
Comment (kosmikus) [reply src/Generics/SOP/Arbitrary.hs:65]: Will you change / simplify this? I still don't completely understand this. Why does it need to be different in behaviour from `shrink` for lists in the example you provide?

**Review by kosmikus — COMMENTED** [review]
You seem to have indicated yourself that you would like to simplify `gshrink` somewhat. I'd also appreciate if `gshrink'` itself was documented. But in principle, I'm happy to merge this.

File: src/Generics/SOP/Arbitrary.hs (hunk 2 above)
Comment (phadej) [reply src/Generics/SOP/Arbitrary.hs:65]: `shrink` in `QuickCheck` provides subterms also: http://hackage.haskell.org/package/QuickCheck-2.9.2/docs/src/Test-QuickCheck-Arbitrary.html#subterms

I guess I can implement something like that with `generics-sop` too

**Pull request closed without merging**